<!DOCTYPE HTML>
<html>
<head>
  <meta charset="utf-8">
  
  <title>PHP中的错误处理、异常处理机制 | 博客</title>

  
  <meta name="author" content="403 Forbidden">
  

  
  <meta name="description" content="在编写php程序时，错误处理是一个重要的部分。如果程序中缺少错误检测代码，那么看上去很不专业，也为安全风险敞开了大门
例：12
那么正确的写法应该如下：12
一、PHP错误处理的三种方式
A、简单的die()语句；    等价于exit();例：1234567if(!file_exists(&amp;apo">
  

  
  
  <meta name="keywords" content="">
  

  <meta id="viewport" name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no, minimal-ui">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">

  <meta property="og:title" content="PHP中的错误处理、异常处理机制"/>

  <meta property="og:site_name" content="博客"/>

  
  <meta property="og:image" content="/favicon.ico"/>
  

  <link href="/favicon.ico" rel="icon">
  <link rel="alternate" href="/atom.xml" title="博客" type="application/atom+xml">
  <link rel="stylesheet" href="/css/style.css" media="screen" type="text/css">
</head>


<body>
<div class="blog">
  <div class="content">

    <header>
  <div class="site-branding">
    <h1 class="site-title">
      <a href="/">博客</a>
    </h1>
    <p class="site-description"></p>
  </div>
  <nav class="site-navigation">
    <ul>
      
        <li><a href="/">主页</a></li>
      
        <li><a href="/archives">归档</a></li>
      
    </ul>
  </nav>
</header>

    <main class="site-main posts-loop">
    <article>

  
    
    <h3 class="article-title"><span>PHP中的错误处理、异常处理机制</span></h3>
    
  

  <div class="article-top-meta">
    <span class="posted-on">
      <a href="/2013/06/23/PHP中的错误处理、异常处理机制/" rel="bookmark">
        <time class="entry-date published" datetime="2013-06-23T05:49:42.000Z">
          2013-06-23
        </time>
      </a>
    </span>
  </div>


  

  <div class="article-content">
    <div class="entry">
      
        <p>在编写php程序时，错误处理是一个重要的部分。如果程序中缺少错误检测代码，那么看上去很不专业，也为安全风险敞开了大门</p>
<p>例：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure></p>
<p>那么正确的写法应该如下：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure></p>
<p>一、PHP错误处理的三种方式</p>
<p>A、简单的die()语句；<br>    等价于exit();<br>例：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">if(!file_exists(&apos;aa.txt&apos;))&#123;</span><br><span class="line">	die(&apos;文件不存在&apos;);</span><br><span class="line">&#125; else &#123;</span><br><span class="line">	//执行操作</span><br><span class="line">&#125;</span><br><span class="line">//如果上面die()被触发，那么这里echo接不被执行</span><br><span class="line">echo &apos;ok&apos;;</span><br></pre></td></tr></table></figure></p>
<pre><code>简洁写法：
</code></pre><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">file_exits(&apos;aaa.txt&apos;) or die(&apos;文件不存在&apos;);</span><br><span class="line">echo &apos;ok&apos;;</span><br></pre></td></tr></table></figure>
<p>B、自定义错误和错误触发器</p>
<p>1、错误处理器（自定义错误，一般用于语法错误处理）<br>        创建自定义错误函数（处理器），该函数必须有能力处理至少两个参数（error_level和errormessage），但是可以接受最多五个参数（error_file、error_line、error_context）<br>        语法：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">function error_function($error_level,$error_message,$error_file,$error_line,$error_context)</span><br><span class="line">//创建好后还需要改写set_error_handler();函数</span><br><span class="line">set_error_handler(&apos;error_function&apos;,E_WARNING);//这里error_function对应上面创建的自定义处理器名，第二个参数为使用自定义错误处理器的错误级别；</span><br></pre></td></tr></table></figure></p>
<p>错误报告级别(了解即可)</p>
<p>这些错误报告级别是错误处理程序旨在处理的错误的不同的类型：</p>
<table class="dataintable"><br><tbody><br><tr>值常量描述</tr><br><tr><br><br><td>2<br></td><br><br><td>E_WARNING<br></td><br><br><td>非致命的 run-time 错误。不暂停脚本执行。<br></td><br></tr><br><tr><br><br><td>8<br></td><br><br><td>E_NOTICE<br></td><br><br><td><br><br><br>Run-time 通知。<br><br><br><br><br>脚本发现可能有错误发生，但也可能在脚本正常运行时发生。<br><br><br><br></td><br></tr><br><tr><br><br><td>256<br></td><br><br><td>E_USER_ERROR<br></td><br><br><td>致命的用户生成的错误。这类似于程序员使用 PHP 函数 trigger_error() 设置的 E_ERROR。<br></td><br></tr><br><tr><br><br><td>512<br></td><br><br><td>E_USER_WARNING<br></td><br><br><td>非致命的用户生成的警告。这类似于程序员使用 PHP 函数 trigger_error() 设置的 E_WARNING。<br></td><br></tr><br><tr><br><br><td>1024<br></td><br><br><td>E_USER_NOTICE<br></td><br><br><td>用户生成的通知。这类似于程序员使用 PHP 函数 trigger_error() 设置的 E_NOTICE。<br></td><br></tr><br><tr><br><br><td>4096<br></td><br><br><td>E_RECOVERABLE_ERROR<br></td><br><br><td>可捕获的致命错误。类似 E_ERROR，但可被用户定义的处理程序捕获。(参见 set_error_handler())<br></td><br></tr><br><tr><br><br><td>8191<br></td><br><br><td>E_ALL<br></td><br><br><td><br><br><br>所有错误和警告，除级别 E_STRICT 以外。<br><br><br><br><br>（在 PHP 6.0，E_STRICT 是 E_ALL 的一部分）<br><br><br><br><br><br><br></td><br></tr><br></tbody><br></table>

<p>2、错误触发器（一般用于处理逻辑上的错误）<br>需求：比如要接收一个年龄，如果数字大于120，就认为是一个错误<br>传统方法：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">if($age&gt;120)&#123;</span><br><span class="line">	echo &apos;年龄错误&apos;;exit();</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure></p>
<pre><code>使用触发器：
</code></pre><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">if($age&gt;120)&#123;</span><br><span class="line">	//trigger_error(&apos;错误信息&apos;[，&apos;错误等级&apos;]);这里错误等级为可选项，用于定义该错误的级别</span><br><span class="line">	//用户定义的级别包含以下三种：E_USER_WARNING 、E_USER_ERROR 、E_USER_NOTICE</span><br><span class="line">	trigger_error(&apos;年龄错误&apos;);//这里是调用的系统默认的错误处理方式，我们也可以用自定义处理器</span><br><span class="line">&#125;</span><br><span class="line">//自定义处理器，与上面相同</span><br><span class="line">function myerror($error_level,$error_message)&#123;</span><br><span class="line">	echo &apos;error text&apos;;</span><br><span class="line">&#125;</span><br><span class="line">//同时需要改变系统默认的处理函数</span><br><span class="line">set_error_handler(&apos;myerror&apos;,E_USER_WARNING);//同上面，第一个参数为自定义函数的名称，第二个为错误级别【这里的错误级别通常为以下三种：E_USER_WARNING 、E_USER_ERROR 、E_USER_NOTICE】</span><br><span class="line">//现在再使用trigger_error就可以使用自定义的错误处理函数了</span><br></pre></td></tr></table></figure>
<p>练习题：<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure></p>
<p>C、错误日志 </p>
<pre><code>默认的根据php.ini中error_log配置，php向服务器的错误记录系统或文件发送错误记录。通过使用error_log()函数可以向文件或远程目的地发送错误记录;
语法：
</code></pre><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">error_log(error[,type,destination,headers])</span><br></pre></td></tr></table></figure>
<pre><code>type部分一般用3，表示在文件后面追加错误信息，而不会覆盖原内容
destination表示目的地，即存放的文件或远程目的地
如：<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">```</span><br><span class="line"></span><br><span class="line"> </span><br><span class="line">二、PHP异常处理【重点】</span><br><span class="line"></span><br><span class="line">1、基本语法</span><br></pre></td></tr></table></figure>

    try{
        //可能出现错误或异常的代码
        //catch 捕获  Exception是php已定义好的异常类
    } catch(Exception $e){
        //对异常处理，方法：
            //1、自己处理
            //2、不处理，将其再次抛出
    }
</code></pre><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"></span><br><span class="line">2、处理程序应当包括：</span><br><span class="line">Try - 使用异常的函数应该位于 &quot;try&quot;  代码块内。如果没有触发异常，则代码将照常继续执行。但是如果异常被触发，会抛出一个异常。</span><br><span class="line">Throw - 这里规定如何触发异常。每一个 &quot;throw&quot; 必须对应至少一个 &quot;catch&quot;</span><br><span class="line">Catch - &quot;catch&quot; 代码块会捕获异常，并创建一个包含异常信息的对象 </span><br><span class="line">让我们触发一个异常：</span><br></pre></td></tr></table></figure>
<p>1){<br>         throw new Exception(“Value must be 1 or below”);<br>     }<br>     return true;<br> }</p>
<p> //在 “try” 代码块中触发异常<br> try{<br>     checkNum(2);<br>     //如果异常被抛出，那么下面一行代码将不会被输出<br>     echo ‘If you see this, the number is 1 or below’;<br> }catch(Exception $e){<br>     //捕获异常<br>     echo ‘Message: ‘ .$e-&gt;getMessage();<br> }<br> ?&gt;</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"></span><br><span class="line">上面代码将获得类似这样一个错误：</span><br><span class="line"></span><br><span class="line">Message: Value must be 1 or below </span><br><span class="line">例子解释：</span><br><span class="line"></span><br><span class="line">上面的代码抛出了一个异常，并捕获了它：</span><br><span class="line"></span><br><span class="line">创建 checkNum() 函数。它检测数字是否大于 1。如果是，则抛出一个异常。</span><br><span class="line">在 &quot;try&quot; 代码块中调用 checkNum() 函数。</span><br><span class="line">checkNum() 函数中的异常被抛出</span><br><span class="line">&quot;catch&quot; 代码块接收到该异常，并创建一个包含异常信息的对象 ($e)。</span><br><span class="line">通过从这个 exception 对象调用 $e-&gt;getMessage()，输出来自该异常的错误消息</span><br><span class="line">不过，为了遵循“每个 throw 必须对应一个 catch”的原则，可以设置一个顶层的异常处理器来处理漏掉的错误。</span><br></pre></td></tr></table></figure>
<p>set_exception_handler()函数可设置处理所有未捕获异常的用户定义函数</p>
<p>//设置一个顶级异常处理器</p>
<p>function myexception($e){</p>
<p>　　 echo ‘this is top exception’;</p>
<p>} //修改默认的异常处理器</p>
<p>set_exception_handler(“myexception”);</p>
<p>try{</p>
<p>　　$i=5;</p>
<p>　　if($i&lt;10){</p>
<p>　　　　throw new exception(‘$i must greater than 10’);</p>
<p>　　}</p>
<p>}catch(Exception $e){</p>
<p>　　//处理异常</p>
<p>　　echo $e-&gt;getMessage().’<br>‘;</p>
<p>　　//不处理异常，继续抛出</p>
<p>　　throw new exception(‘errorinfo’); //也可以用throw $e 保留原错误信息;</p>
<p>}</p>
<p>创建一个自定义的异常类</p>
<p>class customException extends Exception{</p>
<p>　　public function errorMessage(){</p>
<p>　　　　//error message $errorMsg = ‘Error on line ‘.$this-&gt;getLine().’ in ‘.$this-&gt;getFile().’: <strong>‘.$this-&gt;getMessage().’</strong> is not a valid E-Mail address’; return $errorMsg;</p>
<p>　　}</p>
<p>}</p>
<p>//使用</p>
<p>try{</p>
<p>　　throw new customException(‘error message’);</p>
<p>}catch(customException $e){</p>
<p>　　echo $e-&gt;errorMsg();</p>
<p>}</p>
<p>可以使用多个catch来返回不同情况下的错误信息</p>
<p>try{</p>
<p>　　$i=5;</p>
<p>　　if($i&gt;0){</p>
<p>　　　　throw new customException(‘error message’);//使用自定义异常类处理</p>
<p>　　} if($i&lt;-10){</p>
<p>　　　　throw new exception(‘error2’);//使用系统默认异常处理</p>
<p>　　}</p>
<p>}catch(customException $e){</p>
<p>　　echo $e-&gt;getMessage();</p>
<p>}catch(Exception $e1){</p>
<p>　　echo $e1-&gt;getMessage();</p>
<p>}</p>
<p><code>`</code></p>
<p>异常的规则</p>
<ol>
<li>需要进行异常处理的代码应该放入 try 代码块内，以便捕获潜在的异常。</li>
<li>每个try或throw代码块必须至少拥有一个对应的 catch 代码块。</li>
<li>使用多个 catch 代码块可以捕获不同种类的异常。</li>
<li>可以在try代码内的catch 代码块中再次抛出（re-thrown）异常。</li>
<li>简而言之：如果抛出了异常，就必须捕获它。</li>
</ol>

      
    </div>

  </div>

  <div class="article-footer">
    <div class="article-meta pull-left">

    
      

    <span class="post-categories">
      <i class="icon-categories"></i>
        <a href="/categories/Web-开发/">Web 开发</a>
    </span>
    

    

    </div>

    
  </div>
</article>


    </main>

    <footer class="site-footer">
  <p class="site-info">
    Proudly powered by <a href="https://hexo.io/" target="_blank">Hexo</a> and
    Theme by <a href="https://github.com/CodeDaraW/Hacker" target="_blank">Hacker</a>
    </br>
    
    &copy; 2018 403 Forbidden
    
  </p>
</footer>
    
  </div>
</div>
</body>
</html>